README.md
# DART WIF

Bitcoin Wallet Import Format encoding/decoding module.

## Example

``` dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_wif/dart_wif.dart';

void main() {
  Uint8List privateKey = Uint8List.fromList(hex.decode(
      '0000000000000000000000000000000000000000000000000000000000000001'));
  final WIF decoded =
      WIF(version: 128, privateKey: privateKey, compressed: true);
  String key = wif.encode(decoded); // for the testnet use: Wif.encode(239, ...
  print(key);
  // => KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU73sVHnoWn

  var obj = wif.decode(key);

  print(obj);
  // => {
  //	version: 128,
  //	privateKey: "0000000000000000000000000000000000000000000000000000000000000001",
  //	compressed: true
  //}

  try {
    wif.decode(key, 0x09);
  } catch (e) {
    print(e);
  }
  // => Invalid argument(s): Invalid network version
}
```

## LICENSE [MIT](LICENSE)

CHANGELOG.md
## 2.1.1

- update dart_bs58check package version

## 2.1.0

- update dart_bs58check package

## 2.0.0

- update for Codec and Converters

## 1.1.0

- update dart_bs58check package

## 1.0.2

- update pubspec.yaml description

## 1.0.1

- update readme header
- update description

## 1.0.0

- Initial version.


test/fixtures.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';

enum FixtureEnum { valid, invalidEncode, invalidDecode }

class Fixture {
  String? wif;

  bool? compressed;

  Uint8List? get privateKey {
    if (privateKeyHex != null) {
      return Uint8List.fromList(hex.decode(privateKeyHex!));
    }
  }

  String? privateKeyHex;

  int? version;

  String? exception;

  Fixture(
      {this.wif,
      this.exception,
      this.compressed,
      this.privateKeyHex,
      this.version});
}

Map<FixtureEnum, List<Fixture>> fixtures = {
  FixtureEnum.valid: [
    Fixture(
        wif: "KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU73sVHnoWn",
        compressed: true,
        privateKeyHex:
            "0000000000000000000000000000000000000000000000000000000000000001",
        version: 128),
    Fixture(
        wif: "5HpHagT65TZzG1PH3CSu63k8DbpvD8s5ip4nEB3kEsreAnchuDf",
        compressed: false,
        privateKeyHex:
            "0000000000000000000000000000000000000000000000000000000000000001",
        version: 128),
    Fixture(
        wif: "KxhEDBQyyEFymvfJD96q8stMbJMbZUb6D1PmXqBWZDU2WvbvVs9o",
        compressed: true,
        privateKeyHex:
            "2bfe58ab6d9fd575bdc3a624e4825dd2b375d64ac033fbc46ea79dbab4f69a3e",
        version: 128),
    Fixture(
        wif: "KzrA86mCVMGWnLGBQu9yzQa32qbxb5dvSK4XhyjjGAWSBKYX4rHx",
        compressed: true,
        privateKeyHex:
            "6c4313b03f2e7324d75e642f0ab81b734b724e13fec930f309e222470236d66b",
        version: 128),
    Fixture(
        wif: "5JdxzLtFPHNe7CAL8EBC6krdFv9pwPoRo4e3syMZEQT9srmK8hh",
        compressed: false,
        privateKeyHex:
            "6c4313b03f2e7324d75e642f0ab81b734b724e13fec930f309e222470236d66b",
        version: 128),
    Fixture(
        wif: "cRD9b1m3vQxmwmjSoJy7Mj56f4uNFXjcWMCzpQCEmHASS4edEwXv",
        compressed: true,
        privateKeyHex:
            "6c4313b03f2e7324d75e642f0ab81b734b724e13fec930f309e222470236d66b",
        version: 239),
    Fixture(
        wif: "92Qba5hnyWSn5Ffcka56yMQauaWY6ZLd91Vzxbi4a9CCetaHtYj",
        compressed: false,
        privateKeyHex:
            "6c4313b03f2e7324d75e642f0ab81b734b724e13fec930f309e222470236d66b",
        version: 239),
    Fixture(
        wif: "L5oLkpV3aqBjhki6LmvChTCV6odsp4SXM6FfU2Gppt5kFLaHLuZ9",
        compressed: true,
        privateKeyHex:
            "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
        version: 128)
  ],
  FixtureEnum.invalidEncode: [
    Fixture(
        exception: "Invalid privateKey length",
        version: 128,
        privateKeyHex: "fffffffffffeba")
  ],
  FixtureEnum.invalidDecode: [
    Fixture(
        exception: "Invalid network version",
        version: 128,
        wif: "92Qba5hnyWSn5Ffcka56yMQauaWY6ZLd91Vzxbi4a9CCetaHtYj"),
    Fixture(
        exception: "Invalid compression flag",
        version: 128,
        wif: "KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU73sfZr2ym"),
    Fixture(
        exception: "Invalid WIF length",
        version: 128,
        wif: "3tq8Vmhh9SN5XhjTGSWgx8iKk59XbKG6UH4oqpejRuJhfYD"),
    Fixture(
        exception: "Invalid WIF length",
        version: 128,
        wif: "38uMpGARR2BJy5p4dNFKYg9UsWNoBtkpbdrXDjmfvz8krCtw3T1W92ZDSR")
  ]
};


test/dart_wif_test.dart
import 'package:convert/convert.dart';
import 'package:dart_wif/dart_wif.dart';
import 'package:test/test.dart';

import 'fixtures.dart';

void main() {
  if (fixtures.containsKey(FixtureEnum.valid)) {
    for (Fixture fixture in fixtures[FixtureEnum.valid]!) {
      test(
          'encode/encodeRaw returns ${fixture.wif!} for ${fixture.privateKeyHex!.substring(0, 20)}... (${fixture.version!.toString()})',
          () {
        final WIF decoded = WIF(
            version: fixture.version!,
            privateKey: fixture.privateKey!,
            compressed: fixture.compressed!);
        String actual = wif.encode(decoded);
        expect(actual, equals(fixture.wif));
      });
      test(
          'decode/decodeRaw returns ${fixture.privateKeyHex!.substring(0, 20)}... (${fixture.version!}) for ${fixture.wif!}',
          () {
        WIF actual = wif.decode(fixture.wif!, fixture.version);
        expect(actual.version, equals(fixture.version));
        expect(hex.encode(actual.privateKey), equals(fixture.privateKeyHex));
        expect(actual.compressed, equals(fixture.compressed));
      });

      test('decode/encode for ${fixture.wif}', () {
        final decoded = WIF(
            version: fixture.version!,
            privateKey: fixture.privateKey!,
            compressed: fixture.compressed!);
        final encoded = wif.encode(decoded);

        WIF actual = wif.decode(encoded, fixture.version);

        expect(actual, equals(actual));
      });
    }
  }
  if (fixtures.containsKey(FixtureEnum.invalidDecode)) {
    for (Fixture fixture in fixtures[FixtureEnum.invalidDecode]!) {
      test("throws ${fixture.exception} for ${fixture.wif}", () {
        WIF? buffer;
        try {
          buffer = wif.decode(fixture.wif!, fixture.version);
        } catch (err) {
          expect((err as ArgumentError).message, fixture.exception);
        } finally {
          expect(buffer, null);
        }
      });
    }
  }
  if (fixtures.containsKey(FixtureEnum.invalidEncode)) {
    for (Fixture fixture in fixtures[FixtureEnum.invalidEncode]!) {
      test("throws ${fixture.exception} for ${fixture.privateKeyHex}", () {
        String? buffer;
        final WIF decoded =
            WIF(privateKey: fixture.privateKey!, version: fixture.version!);

        try {
          buffer = wif.encode(decoded);
        } catch (err) {
          expect((err as ArgumentError).message, fixture.exception);
        } finally {
          expect(buffer, null);
        }
      });
    }
  }
}


lib/src/models/wif.dart
import 'dart:typed_data';

/// WIF
class WIF {
  /// Version
  int version;

  /// Private Key
  Uint8List privateKey;

  /// Is Compressed WIF ?
  bool compressed;

  WIF(
      {required this.version,
      required this.privateKey,
      this.compressed = false});

  /// equal operator
  @override
  bool operator ==(covariant WIF other) {
    return (version == other.version &&
        privateKey.map((e) => e.toRadixString(16).padLeft(2, '0')).join('') ==
            privateKey
                .map((e) => e.toRadixString(16).padLeft(2, '0'))
                .join('') &&
        compressed == other.compressed);
  }

  /// Hash code
  @override
  int get hashCode => Object.hashAll([version, privateKey, compressed]);

  /// To String
  @override
  String toString() {
    List<String> str = [
      '"version": "$version"',
      '"privateKey": "${privateKey.map((e) => e.toRadixString(16).padLeft(2, '0')).join('')}"',
      '"compressed": ${compressed.toString()}'
    ];

    return "{${str.join(",")}}";
  }
}


lib/src/converters/wif_decoder.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dart_bs58check/dart_bs58check.dart';
import 'package:dart_wif/dart_wif.dart';

/// The canonical instance of [WIFEncoder].
const wifDecoder = WIFDecoder();

/// The canonical instance of [WIFDecoder].
class WIFDecoder extends Converter<String, WIF> {
  const WIFDecoder();

  /// Convert [String] to [WIF]
  @override
  WIF convert(String input, [int? version]) {
    Uint8List buffer = bs58check.decode(input);
    if (version != null && buffer[0] != version) {
      throw ArgumentError('Invalid network version');
    }

    // uncompressed
    if (buffer.lengthInBytes == 33) {
      return WIF(
          version: buffer[0], privateKey: buffer.sublist(1), compressed: false);
    }

    // invalid length
    if (buffer.length != 34) {
      throw ArgumentError('Invalid WIF length');
    }

    // invalid compression flag
    if (buffer[33] != 0x01) {
      throw ArgumentError('Invalid compression flag');
    }

    // compressed
    return WIF(
        version: buffer[0],
        privateKey: buffer.sublist(1, 33),
        compressed: true);
  }
}


example/dart_wif_example.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_wif/dart_wif.dart';

void main() {
  Uint8List privateKey = Uint8List.fromList(hex.decode(
      '0000000000000000000000000000000000000000000000000000000000000001'));
  final WIF decoded =
      WIF(version: 128, privateKey: privateKey, compressed: true);
  String key = wif.encode(decoded); // for the testnet use: Wif.encode(239, ...
  print(key);
  // => KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU73sVHnoWn

  var obj = wif.decode(key);

  print(obj);
  // => {
  //	version: 128,
  //	privateKey: "0000000000000000000000000000000000000000000000000000000000000001",
  //	compressed: true
  //}

  try {
    wif.decode(key, 0x09);
  } catch (e) {
    print(e);
  }
  // => Invalid argument(s): Invalid network version

  print(obj == decoded);
  // => true

  print(obj == WIF(privateKey: Uint8List(0), version: 1, compressed: false));
  // => false
}
